### PR TITLE
[IPSCCP] Add range attribute handling.

### PR DESCRIPTION
**Pull request by andjo403**

### CODE REVIEW

**Comment by andjo403** [comment]
CC @nikic as you seems to have done changes here also for the range metadata.

**Review by antoniofrighetto — APPROVED** [review]
LGTM, `markConstantRange` implementation looks good. Let's wait for `@nikic` as well when he's back though.

File: llvm/include/llvm/Transforms/Utils/SCCPSolver.h [hunk 1]
```diff
@@ -151,6 +151,10 @@ class SCCPSolver {
   /// works with both scalars and structs.
   void markOverdefined(Value *V);
 
+  /// trackValueOfArgument - Mark the specified argument overdefined unless it
+  /// have range attribute.  This works with both scalars and structs.
```
Comment (dtcxzyw) [inline llvm/include/llvm/Transforms/Utils/SCCPSolver.h:155]: >  This works with both scalars and structs.

I don't understand why it works with structs. Doesn't it only work with integer scalars?

File: llvm/lib/Transforms/Utils/SCCPSolver.cpp [hunk 2]
```diff
@@ -788,6 +794,17 @@ class SCCPInstVisitor : public InstVisitor<SCCPInstVisitor> {
       markOverdefined(ValueState[V], V);
   }
 
+  void trackValueOfArgument(Argument *A) {
+    if (A->getType()->isIntegerTy()) {
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/SCCPSolver.cpp:798]: Do we need the restriction to non-vectors here? IIRC ipsccp can handle vectors though it mostly doesn't.

File: llvm/test/Transforms/SCCP/range-attribute.ll [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  %c1 = icmp ult i32 %v, 10
+  call void @use(i1 %c1)
+  %c2 = icmp ult i32 %v, 9
+  call void @use(i1 %c2)
+  %c3 = icmp ugt i32 %v, 9
+  call void @use(i1 %c3)
+  %c4 = icmp ugt i32 %v, 8
+  call void @use(i1 %c4)
+  ret void
+}
```
Comment (nikic) [inline llvm/test/Transforms/SCCP/range-attribute.ll:82]: Can you please also check that range() does not pessimize inter-procedural inference? I.e. if we get better info from call-sites, we should use it.

File: llvm/include/llvm/Transforms/Utils/SCCPSolver.h (hunk 1 above)
Comment (andjo403) [reply llvm/include/llvm/Transforms/Utils/SCCPSolver.h:155]: as it calls the markOverdefined if it is not a integer with a range attribute and that function have that text I selected to have it for this function also.

File: llvm/lib/Transforms/Utils/SCCPSolver.cpp (hunk 2 above)
Comment (andjo403) [reply llvm/lib/Transforms/Utils/SCCPSolver.cpp:798]: I only assumed that vectors was not handled as there was a restriction to only non-vectors for the range metadata.

File: llvm/test/Transforms/SCCP/range-attribute.ll (hunk 3 above)
Comment (andjo403) [reply llvm/test/Transforms/SCCP/range-attribute.ll:82]: added test for this.

File: llvm/lib/Transforms/Utils/SCCPSolver.cpp (hunk 2 above)
Comment (nikic) [reply llvm/lib/Transforms/Utils/SCCPSolver.cpp:798]: Ok, we can relax this separately from this PR.

File: llvm/lib/Transforms/Utils/SCCPSolver.cpp [hunk 4]
```diff
@@ -873,6 +890,15 @@ bool SCCPInstVisitor::markConstant(ValueLatticeElement &IV, Value *V,
   return true;
 }
 
+bool SCCPInstVisitor::markConstantRange(ValueLatticeElement &IV, Value *V,
+                                        ConstantRange &CR) {
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/SCCPSolver.cpp:894]: Should pass by const reference here (or by value and use std::move, but that's not compatible with the dbg printing).

File: llvm/test/Transforms/SCCP/range-attribute.ll (hunk 3 above)
Comment (nikic) [reply llvm/test/Transforms/SCCP/range-attribute.ll:82]: This isn't quite what I was looking for, or maybe I'm misunderstanding the test. What I had in mind if a function with range(0, 10) that has icmp 5 and is only called with value 5, or something like that.

Comment (andjo403) [reply llvm/test/Transforms/SCCP/range-attribute.ll:82]: ok I thought that you was thinking of the test ip_load_range that exist for range metadata
https://github.com/andjo403/llvm-project/blob/fff4a199f2b1286618ee129726a71e753c0b7ea1/llvm/test/Transforms/SCCP/metadata.ll#L102-L119

So I think I did kind of the opposite of the test you was thinking of, that the range is propagated in to a function without a range and you want a constant to be propagated in to a function with a argument that have a range if I understand correctly.

File: llvm/lib/Transforms/Utils/SCCPSolver.cpp (hunk 4 above)
Comment (andjo403) [reply llvm/lib/Transforms/Utils/SCCPSolver.cpp]: added const

File: llvm/lib/Transforms/Utils/SCCPSolver.cpp (hunk 2 above)
Comment (andjo403) [reply llvm/lib/Transforms/Utils/SCCPSolver.cpp:799]: Tried to add a test for vector like below but was not optimized when I remove the check. But maybe I did something wrong.
```
define void @range_attribute_vec(<2 x i32> range(i32 0, 10) %v) {
  %c1 = icmp ult <2 x i32> %v, <i32 10, i32 10>
  call void @use(<2 x i1> %c1)
  %c2 = icmp ult <2 x i32> %v, <i32 9, i32 9>
  call void @use(<2 x i1> %c2)
  %c3 = icmp ugt <2 x i32> %v, <i32 9, i32 9>
  call void @use(<2 x i1> %c3)
  %c4 = icmp ugt <2 x i32> %v, <i32 8, i32 8>
  call void @use(<2 x i1> %c4)
  ret void
}
```

File: llvm/test/Transforms/SCCP/range-attribute.ll (hunk 3 above)
Comment (andjo403) [reply llvm/test/Transforms/SCCP/range-attribute.ll:82]: think that my test fails as ConstantVector is not handled in https://github.com/andjo403/llvm-project/blob/fa71860e1937b916758693e268d5a16021c8b5c3/llvm/include/llvm/Analysis/ValueLattice.h#L301-L339

File: llvm/lib/Transforms/Utils/SCCPSolver.cpp (hunk 2 above)
Comment (andjo403) [reply llvm/lib/Transforms/Utils/SCCPSolver.cpp:799]: think that my test fails as ConstantVector is not handled in https://github.com/andjo403/llvm-project/blob/fa71860e1937b916758693e268d5a16021c8b5c3/llvm/include/llvm/Analysis/ValueLattice.h#L301-L339

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**